hoomd/mpcd/CosineChannelGeometry.h

```diff
@@ -0,0 +1,317 @@
+            }
+
+        /* Calculate position (x0,y0,z0) of collision with wall:
+         *  Because there is no analytical solution for f(x) = cos(x)-x = 0, we use Newton's method
+         * to numerically estimate the x positon of the intersection first. It is convinient to use
+         * the halfway point between the last particle position outside the wall (at time t-dt) and
+         * the current position inside the wall (at time t) as initial guess for the intersection.
+         *
+         *  We limit the number of iterations (max_iteration) and the desired presicion
+         * (target_precision) for performance reasons.
+         */
+        const unsigned int max_iteration = 6;
+        const Scalar target_precision = 1e-5;
+
+        Scalar x0 = pos.x - 0.5 * dt * vel.x;
+        Scalar y0;
+        Scalar z0;
+
+        /* catch the case where a particle collides exactly vertically (v_x=0 -> old x pos = new x
+         * pos) In this case, y0 = -(0)*0/0 + (y-dt*v_y) == nan, should be y0 =(y-dt*v_y)
+         */
+        if (vel.x == 0) // exactly vertical x-collision
+            {
+            x0 = pos.x;
+            y0 = (m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H);
+            z0 = (pos.z - dt * vel.z);
+            }
+        else if (vel.y == 0) // exactly horizontal z-collision
+            {
+            x0 = 1 / m_wavenumber * fast::acos((pos.y - sign * m_H) / m_amplitude);
+            y0 = pos.y;
+            z0 = -(pos.x - dt * vel.x - x0) * vel.z / vel.x + (pos.z - dt * vel.z);
+            }
+        else
+            {
+            Scalar delta = fabs(0
+                                - ((m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H)
+                                   - vel.y / vel.x * (x0 - pos.x) - pos.y));
+
+            Scalar n, n2;
+            Scalar s, c;
+            unsigned int counter = 0;
+            while (delta > target_precision && counter < max_iteration)
+                {
+                fast::sincos(x0 * m_wavenumber, s, c);
+                n = (m_amplitude * c + sign * m_H) - vel.y / vel.x * (x0 - pos.x) - pos.y; // f
+                n2 = -m_wavenumber * m_amplitude * s - vel.y / vel.x;                      // df
+                x0 = x0 - n / n2; // x = x - f/df
+                delta = fabs(0
+                             - ((m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H)
+                                - vel.y / vel.x * (x0 - pos.x) - pos.y));
+                ++counter;
+                }
+
+            /* The new z position is calculated from the wall equation to guarantee that the new
+             * particle positon is exactly at the wall and not accidentally slightly inside of the
+             * wall because of nummerical presicion.
+             */
+            y0 = (m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H);
+
+            /* The new y position can be calculated from the fact that the last position outside of
```

Comment:
Please update these comments to match the new coordinate system. They should be y and z, not z and y, I think.